stellar-contract-env-host/src/host/validity.rs

```diff
@@ -0,0 +1,67 @@
+use crate::xdr::{ScHostFnErrorCode, ScHostObjErrorCode};
+use crate::{Host, HostError};
+
+impl Host {
+    pub(crate) fn validate_index_lt_bound(
+        &self,
+        index: u32,
+        bound: usize,
+    ) -> Result<(), HostError> {
+        if index as usize >= bound {
+            return Err(self.err_status_msg(
+                ScHostObjErrorCode::VecIndexOutOfBound, // TODO: need to reconcile between InputArgsInvalid and VecIndexOutOfBound
+                "start index out of bound",
+            ));
+        }
+        Ok(())
+    }
+
+    pub(crate) fn validate_index_le_bound(
+        &self,
+        index: u32,
+        bound: usize,
+    ) -> Result<(), HostError> {
+        if index as usize > bound {
+            return Err(self.err_status_msg(
+                ScHostObjErrorCode::VecIndexOutOfBound, // TODO: need to reconcile between InputArgsInvalid and VecIndexOutOfBound
+                "start index out of bound",
+            ));
+        }
+        Ok(())
+    }
+
+    pub(crate) fn validate_start_end_bound(
+        &self,
+        start: u32,
+        end: u32,
+        bound: usize,
+    ) -> Result<(), HostError> {
+        if start as usize >= bound {
+            return Err(self.err_status_msg(
+                ScHostObjErrorCode::VecIndexOutOfBound,
+                "start index out of bound",
+            ));
+        }
+        if end as usize > bound {
+            return Err(self.err_status_msg(
+                ScHostObjErrorCode::VecIndexOutOfBound,
+                "end index out of bound",
+            ));
+        }
+        // we don't check start <= end, as [start..end] is ok in that case
```

Comment:
Is this right? Doesn't `[start..end]` where `start > end` result in a panic?

Reply:
Thanks for the catch. You are right it does cause a panic. The im-rc had additional internal logic that just returns empty Vector if `start > end`, which tricked me to thinking that was ok in general. So glad you caught this. Fixed.